library/ssl_msg.c

```diff
@@ -3106,32 +3106,127 @@ static size_t ssl_get_reassembly_buffer_size( size_t msg_len,
 
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
 
+#if defined(MBEDTLS_SSL_HS_DEFRAG_CLT)
+#define TRY_READ_FLAG_COMMIT 1
+static int ssl_defrag_try_read( mbedtls_ssl_context *ssl, size_t desired, unsigned int flags )
+{
+    int ret, rret;
+    mbedtls_ssl_hs_defrag *defrag = &ssl->handshake->defrag;
+
+    /* Attempt to get the required portion of a handshake message from the reader */
+    ret = mbedtls_reader_get( defrag->reader, desired, &defrag->pmsg, NULL );
+
+    MBEDTLS_SSL_DEBUG_RET( 1, ( "mbedtls_reader_get" ), ret );
+
+    if( ret == 0 && ( flags & TRY_READ_FLAG_COMMIT ) )
+        mbedtls_reader_commit( defrag->reader );
+    else
+    {
+        if( ret == MBEDTLS_ERR_READER_OUT_OF_DATA )
+        {
+            /* This error code indicates that the reader does not
+             * contain sufficient data for the handshake message.
+             *
+             * For example, this will happen if the header of
+             * the handshake message has been split between two
+             * TLS records.
+             *
+             * In this situation, we need to continue to read additional
+             * TLS records and feed their contents to the reader.
+             *
+             * We accomplish this by re-routing the error code into
+             * `MBEDTLS_ERR_SSL_CONTINUE_PROCESSING` and looping.
+             */
+            MBEDTLS_SSL_DEBUG_MSG( 1, (
+                        "Handshake reader needs more data: desired: %d",
+                        desired ) );
+
+            ret = MBEDTLS_ERR_SSL_CONTINUE_PROCESSING;
+        }
+
+        rret = mbedtls_reader_reclaim( defrag->reader, NULL );
```

Comment:
I was expecting this to happen in `ssl_consume_current_message()` - is there a reason to reclaim already here?

Reply:
We have discussed this offline. Adding the summary of our discussion:

The above approach was an attempt to read the header and the body of the message in two takes. Consider a situation where the header and the body of the handshake message reside in two subsequent messages, e.g.

```
TLS record 1               TLS record 2 
         [ hs hdr |            hs body ]
```

However, after discussing this approach offline, we came to conclusion that `reclaim` should only be invoked in the context of `ssl_consume_current_message`. The design document is updated with the details.